Alternative sortByPositionPresorted() and String generation optimization for TheoreticalSpectrumGenerator

akriese opened the pull request.

Commit:
Alternative sortByPosition() and added String generation optimization

Commit:
further work on specialSortByPosition()

Commit:
TheoreticalSpectrumGenerator.cpp

Commit:
Use merge sorting for specialSortByPosition()

Commit:
Use std::move() in MSSpectrum::select()

akriese committed:
Merge branch 'test_branch_2' into chen

akriese committed:
Merge pull request #2 from xchen98/chen

TheoreticalSpectrumGenerator.cpp

Commit:
Resolve earlier merge conflicts, sort unsorted chunks

Commit:
Rename specialSortByPosition() to sortByPositionPresorted(), refactor code

akriese commented:
We've written a new sort-Function for partially sorted chunks of Peaks in a spectrum. Also MSSpectrum::select() got optimized by applying moves and reserves. 
Furthermore, the functions of TheoreticalSpectrumGenerator were optimized by caching Strings and using emplace_back() and reserve() for meta information vectors.
So far, the new sort-Function is only applied for the getSpectrum() function, but might be used to replace stable_sort() in other functions too, due to the fact, that partially sorted chunks exist very often.
The above optimizations speed up getSpectrum() by about 30% in its usage in the SimpleSearchEngine.

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -286,6 +286,8 @@ namespace OpenMS
     */
     void sortByPosition();
 
+    void sortByPositionPresorted(const std::vector<std::pair<Size, bool>>& chunks);
```
use doxygen to describe what `chunks` is

akriese replied:
Gonna do that!

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -208,108 +209,100 @@ namespace OpenMS
   void TheoreticalSpectrumGenerator::addAbundantImmoniumIons_(PeakSpectrum& spectrum, const AASequence& peptide, DataArrays::StringDataArray& ion_names, DataArrays::IntegerDataArray& charges) const
   {
     Peak1D p;
 
```
add a comment here, that these masses need to stay sorted

akriese replied:
What do you mean by that?

Commit:
refactor code, pulled unnecessary String generation out of loops

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -473,16 +485,15 @@ namespace OpenMS
           if (add_metainfo_)
           {
-            ion_names.push_back(ion_name);
+            ion_names.emplace_back(residue_str);
```
this creates the same string in every iteration, right?
its not straight forward to avoid, I guess..

akriese replied:
Yes, because StringDataArray is supposed to have the same size as the ContainerType, I guess.

akriese replied:
Ah I misunderstood the question. I actually fixed that in the latest commit.

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -504,8 +517,17 @@ namespace OpenMS
+                                               DataArrays::StringDataArray& ion_names,
+                                               DataArrays::IntegerDataArray& charges,
+                                               std::vector<std::pair<Size, bool>>& chunks,
+                                               Residue::ResidueType res_type,
```
can you make `const Residue::ResidueType res_type`? or does this break something? Also in `addLosses_...`

akriese replied:
Should be okay to do that.

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -568,28 +590,32 @@ namespace OpenMS
+            chunks.emplace_back(spectrum.size(), false);
+          }
+
+          Peak1D p;
```
declare `p` outside the loop ?

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -568,28 +590,32 @@ namespace OpenMS
+          Peak1D p;
+          p.setMZ(pos);
+          p.setIntensity(intensity);
+          spectrum.push_back(p);
```
or write a Ctor for Peak1D with `Peak1D(mz, intensity)`; then you can just `spectrum.emplace_back(pos, intensity);`

akriese replied:
The emplace_back version sounds good

akriese replied:
The constructor already exists if I'm not mistaken:
`inline Peak1D(PositionType a, IntensityType b) : position_(a), intensity_(b) {}`

cbielow commented on src/openms/source/KERNEL/MSSpectrum.cpp:
```diff
@@ -317,6 +322,49 @@ namespace OpenMS
+
+  void MSSpectrum::sortByPositionPresorted(const std::vector<std::pair<Size, bool>>& chunks)
+  {
+    if (isSorted()) return;
```
this is never true, is it?

cbielow replied:
unless chunks.size()==1 and the chunk is already sorted

akriese replied:
Let's say, almost never. There might be cases depending on what Ion-Types are chosen.

akriese replied:
I can take it out anyway, because the overhead of the new sort might not be too bad on a sorted vector, compared to the isSorted() on every unsorted vector.

cbielow commented on src/openms/source/KERNEL/MSSpectrum.cpp:
```diff
@@ -317,6 +322,49 @@ namespace OpenMS
+    else
+    {
+      std::vector<Size> select_indices(this->size());
+      for (Size i = 0; i < this->size(); ++i) select_indices[i] = i;
```
`std::iota(select_indices.begin(), select_indices.end(), 0);` ? :)

akriese replied:
alright :)

cbielow commented:
really nice work!

Can you say how much time the String creation saves? Its still a bit inconsistent, because sometimes strings are created even though they will never be used (add_meta == false), or they are created in a loop. So, is String creation still a larger chunk in the flame graph?

akriese commented:
I think, add_metainfo_ is almost always true, so creating the Strings is almost always needed.
--> Otherwise we would have to take the loop apart and do one for add_metainfo_==true and the same thing for add_metainfo_==false
String creation is still a bigger part in the flamegraph, but got reduced. Im gonna test, how much exactly.

akriese commented:
Note: I am currently only testing the SimpleSearchEngine, which doesnt use losses, precursors or abundantImmoniumIons. So, we don't actually test the optimizations in those functions right now.

akriese commented:
Looking at the flamegraph, the biggest difference is, that the String::operator+(String) chunk is basically gone. All other chunks stayed more or less the same. The percentage of addPeaks_ went from 4.5% to 3%.

Commit:
make ResidueType parameters const, directly pass position and intensity to emplace_back

akriese committed:
introduce struct Chunk and Chunks, refactoring

akriese committed:
add constness of ResidueType in function declaration too

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,24 @@ namespace OpenMS
+    };
+
+    struct Chunks {
+      std::vector<Chunk> chunks_;
```
`chunks_` are public here, but named as if private. remove the `_`

cbielow replied:
or make it private and provide a `get` method (this is a bit cleaner because noone from outside can operate on the vector)

akriese replied:
made a get-Function

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,24 @@ namespace OpenMS
+
+    struct Chunks {
+      std::vector<Chunk> chunks_;
+      const MSSpectrum& spec_;
```
`spec_` should be private?

akriese replied:
I guess yes, because otherwise one could change the associated spectrum during getSpectrum() (which wouldnt make sense, but still this can be prevented)

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,24 @@ namespace OpenMS
+    struct Chunk {
+      Size start;
+      Size end;
+      bool isSorted;
```
`is_sorted` naming convention

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -286,6 +304,12 @@ namespace OpenMS
 
+    /**
+      @brief Sort the spectrum, but uses the fact, that certain chunks are presorted
+      @param chunks a chunk represents the end of a sublist of peaks in the spectrum, that is (true) or isn't (false) sorted yet
```
`true` and `false` refer to `Chunk::is_sorted` ?

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -373,16 +375,20 @@ namespace OpenMS
+    String residue_str(Residue::residueTypeToIonLetter(res_type));
+    String ion_ordinal_str(String(ion_ordinal) + "-");
+
     std::vector<double> losses;
```
could use a `losses.reserve(f_losses.size());`

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -373,16 +375,20 @@ namespace OpenMS
+    String ion_ordinal_str(String(ion_ordinal) + "-");
+
     std::vector<double> losses;
     std::vector<String> losses_names;
```
could use a `if (add_metainfo) losses_names.reserve(f_losses.size());`

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -373,16 +375,20 @@ namespace OpenMS
                          bool add_metainfo,
                          int charge)
   {
+    String charge_str((Size)abs(charge), '+');
```
these 3 strings can be const?

cbielow commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -778,8 +799,7 @@ namespace OpenMS
+        ion_names.push_back(ion_name_h2o);
         charges.push_back(charge);
       }
       spectrum.push_back(p);
```
there are still a few places where push_back for Peak1D is used. Replace with emplace_back(mz,int) ?

akriese replied:
Replaced all occurences

cbielow commented on src/openms/source/KERNEL/MSSpectrum.cpp:
```diff
@@ -317,6 +322,49 @@ namespace OpenMS
   }
 
+
+  void MSSpectrum::sortByPositionPresorted(const std::vector<Chunk>& chunks)
```
write a test for this in MSSpectrum_test.cpp

akriese replied:
done

cbielow commented:
in line https://github.com/cbielow/OpenMS/blob/f3bc9a326d89ba348d6f6ba8e63f074c3b3a4281/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L198

the local variables are copied to the spectrum data arrays.
It should be much cheaper to move them (both for move-assignment and move-construction cases).

cbielow commented:
does this really work:
an inner chunk:
https://github.com/cbielow/OpenMS/blob/f3bc9a326d89ba348d6f6ba8e63f074c3b3a4281/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L595
wrapped in an outer chunk:
https://github.com/cbielow/OpenMS/blob/f3bc9a326d89ba348d6f6ba8e63f074c3b3a4281/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L608
??

akriese committed:
refactor struct Chunks, avoid construction of Peak1D instances

akriese committed:
refactor, use moves in getSpectrum()

akriese committed:
Add test for sortByPositionPresorted() in MSSpectrum_test.cpp

akriese commented:
The last commits have sped up getSpectrum() once more. Compared to a build from the develop-Branch, the function is now 40% faster (in the SimpleSearchEngine).

akriese committed:
Cache formula_strings in addLosses_faster_(), optimize EmpiricalFormula::operator<()

akriese commented:
I changed the configuration of SimpleSearchEngine to test more functions, that we had optimized in above commits. EmpiriclaFormula::toString() appears to be very expensive. So I added a static map<EmpiricalFormula, String> to cache earlier computed formula-Strings. Also EmpiricalFormula::operator<() was optimized by reordering comparisons. That too brought a good speed-up.

Depending on the SimpleSearchEngine's settings, the speedup lies between 7% and 20%.

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -334,13 +313,20 @@ namespace OpenMS
+                                                        Int charge,
+                                                        double intensity) const
   {
+    const String ion_name = String(Residue::residueTypeToIonLetter(res_type)) + String(ion.size()) + String((Size)abs(charge), '+');
```
This wasn't added by you. You just moved it up. But is it intended that the charge is represented as pluses regardless of positive or negative charge?

IceFreez3r replied:
If charge is always positive, then why abs(charge)

IceFreez3r replied:
Also is there any reason for moving it up here?

akriese replied:
Why to use abs() is a good question. I'll have a look into it :)
Moving the line up there might have been a result of us constructing the used strings at start of each function at the start of the function. Here, it does not make a difference semantically, but adapts the function order of other functions,

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -351,58 +337,51 @@ namespace OpenMS
       if (add_metainfo_) // one entry per peak
       {
-        ion_names.push_back(ion_name);
+        ion_names.emplace_back(ion_name);
```
I think this doesn't change anything as ion_name is constructed before.

akriese replied:
True, gonna change it back to avoid misunderstandings.

IceFreez3r commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,30 @@ namespace OpenMS
+      Size start;
+      Size end;
+      bool is_sorted;
+      Chunk(Size a, Size b, bool c) : start(a), end(b), is_sorted(c) {}
```
Maybe not a, b, c

akriese replied:
Gonna change that!

IceFreez3r commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,30 @@ namespace OpenMS
+        Chunks(const MSSpectrum& s) : spec_(s) {}
+        void add(bool is_sorted)
+        {
+          chunks_.emplace_back((chunks_.empty() ? 0 : chunks_.back().end), spec_.size(), is_sorted);
```
So a chunk goes from start to exclusive end, right? Doesn't this mean, that chunks[i].end is equal to chunks[i+1].start? So you could probably just save the start. This could ofc reduce readability.

akriese replied:
Yes, there is redundant information in Chunks, but it make the sorting easier as we don't have to add a zero-chunk (which was the case earlier).

IceFreez3r commented on src/openms/source/CHEMISTRY/EmpiricalFormula.cpp:
```diff
@@ -643,16 +643,9 @@ namespace OpenMS
+    if (charge_ != rhs.charge_) return charge_ < rhs.charge_;
 
-    return charge_ < rhs.charge_;
+    return formula_ < rhs.formula_;
```
I guess the old code just mimicked the code in formula_<(...), right?

akriese replied:
Yes! Also, switching the comparison of charges with the comparison of the maps apparently saves a good amount of time :)

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -175,149 +175,128 @@ namespace OpenMS
     if (add_abundant_immonium_ions_)
     {
       addAbundantImmoniumIons_(spectrum, peptide, ion_names, charges);
+      chunks.add(true);
```
How do you know here, that the chunk is ordered?

akriese replied:
Have a look into addAbundantImmoniumIons_() ;) I reordered the if-statements so that the order in the chunk must be sorted.

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -175,149 +175,128 @@ namespace OpenMS
         charges.push_back(1);
       }
-      spectrum.push_back(p);
+      spectrum.emplace_back(70.0656, 1.0);
```
I would add a comment here like:
`//emplace_back(mz, intensity)`

IceFreez3r replied:
Same for the other occurrences

akriese replied:
Good idea! Gonna do it for the first if-statement!

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -599,16 +596,18 @@ namespace OpenMS
-          const AASequence ion = peptide.getPrefix(i);
-          addLosses_(spectrum, ion, ion_names, charges, intensity, res_type, charge);
+          // add loss peaks (slow)
+          Size i = add_first_prefix_ion_ ? 1 : 2;
```
As you did it above:
`Size i = Size(!add_first_prefix_ion) + 1`

akriese replied:
Hm, my solution is kind of hacky. In this case one could also say `i = 2 - add_first_prefix_ion_;` but if think the ternary operator is still the most readable. So, I'm gonna leave my solution for `0 : 1` cases and leave "more complex" ones to the ternary operator.

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -674,28 +685,31 @@ namespace OpenMS
-      }
+        chunks.add(true);
 
-      if (add_losses_ && add_isotopes_) // otherwise losses are already added, see above
```
Why can you remove add_isotopes_ here as a condition?

akriese replied:
As the if-statement is nested into the `if(add_isotopes_ )`, we don't have to check it anymore, as it must already be true.

IceFreez3r commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -761,33 +772,28 @@ namespace OpenMS
         {
-          String ion_name("[M+H]-H2O" + String((Size)abs(charge), '+'));
-          ion_names.push_back(ion_name);
+          ion_names.push_back(ion_name_h2o);
```
std::move as ion_name_doesn't get used again?

akriese replied:
Good idea, don't know if it will make a difference, as SSO will be used for the String (it's only 9+charge long).

IceFreez3r reviewed (commented):
I don't quiet get when you can do `chunks.add(true)` and when only `chunks.add(false)`. For the rest i made some (little) comments, but looks good

cbielow commented on src/tests/class_tests/openms/source/MSSpectrum_test.cpp:
```diff
@@ -811,7 +811,84 @@ START_SECTION((void sortByPosition()))
+  MSSpectrum::IntegerDataArray int_array;
+  std::vector<double> mzs, intensities;
+  MSSpectrum::Chunks chunks(ds);
+  intensities.push_back(19);  mzs.push_back(419.113); float_array.push_back(19);  string_array.push_back("19");  int_array.push_back(19);
```
you could use initializer lists i.e. `std::vector<double> mzs = { 1, 2, 3};` to init the arrays. Less code and more readable?

akriese replied:
Tried to do so, just to find out that you can't apply list initialization on FloatDataArrays, as they have that MetaData member...

akriese replied:
Did it with .assign, now.

cbielow commented on src/tests/class_tests/openms/source/MSSpectrum_test.cpp:
```diff
@@ -811,7 +811,84 @@ START_SECTION((void sortByPosition()))
+    if (mzs[i] < last_added) chunks.add(true);
+    last_added = mzs[i];
+    p.setIntensity(intensities[i]); p.setMZ(mzs[i]);
+    ds.push_back(p);
```
emplace_back(mz, int)?

cbielow commented on src/tests/class_tests/openms/source/MSSpectrum_test.cpp:
```diff
@@ -811,7 +811,84 @@ START_SECTION((void sortByPosition()))
+  MSSpectrum::StringDataArray::iterator it3 = ds.getStringDataArrays()[0].begin();
+  MSSpectrum::IntegerDataArray::iterator it4 = ds.getIntegerDataArrays()[0].begin();
+  std::sort(intensities.begin(), intensities.end());
+  for(std::vector<double>::iterator it = intensities.begin(); it != intensities.end(); ++it)
```
the linter will complain:
use `for (` and `if (`

cbielow commented on src/tests/class_tests/openms/source/MSSpectrum_test.cpp:
```diff
@@ -811,7 +811,84 @@ START_SECTION((void sortByPosition()))
+    }
+    else
+    {
+      TEST_EQUAL(true,false)
```
this is hacky :)
what about
```
ABORT_IF(intensities.size() != ds.getStringDataArrays()[0].size())
.... < more size tests>

for (size_t i = 0; i < intensities.size(); ++i)
{
  TEST_EQUAL(intensities[i], ds.getStringDataArrays()[0][i])
   ... < more tests>
}
```

akriese replied:
Done.

akriese committed:
Added comments, small refactorings

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,30 @@ namespace OpenMS
+    struct Chunk {
+      Size start;
+      Size end;
+      bool is_sorted;
```
annotate with e.g. `bool is_sorted; ///< are the elements in [start, end) sorted already?` or similar

cbielow commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -77,6 +77,30 @@ namespace OpenMS
+    /// Used to remember what subsets in a spectrum are sorted already to allow faster sorting of the spectrum
+    struct Chunk {
+      Size start;
+      Size end;
```
say that `end` index is not inclusive

akriese committed:
Refactor MSSpectrum_test, more comments

akriese committed:
make addLossesFaster_() and formula_str_cache_ members of TSG, use pointers for MetaDataArrays to reduce moving

akriese committed:
Fix weird copy-paste-error

akriese committed:
precompute formula_str_cache

akriese committed:
Add missing omp atomic (doesn't completely fix multithreading)

akriese committed:
make AnnotatedHit_::operator<() unambiguous to make the sorting threadsafe

akriese committed:
sort peptide_ids after post-processing

akriese committed:
use strncmp(), only sort peptide_ids when using more than one thread

akriese committed:
only precompute formula_str_cache if adding losses

akriese committed:
change order of comparisons in hasBetterScore()

akriese committed:
avoid nullptr-dereferencing with metadata-arrays
